### PR TITLE
Communicator messages recipient list filtering

### LINKED ISSUE #5370: Communicator message's recipients filtering
We need to add more filtering to communicator message recipients based on logged in users role.

If message is sent to a students, usergroups and workspaces we need to filter recipient list based on recipients access rights.

Students should only see usergroups and workspace in the recipient list that he/she is part of.

Teachers, managers, study program leaders and admins need to follow the same visibility rules that are present in Guider currently.

### PR DESCRIPTION
**Pull request by saaralepisto**
closes #5370

### CODE REVIEW

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java [hunk 1]
```diff
@@ -350,11 +355,50 @@ public CommunicatorMessageRecipient findCommunicatorMessageRecipientByMessageAnd
...
+    } else if (loggedUser.defaultSchoolDataIdentifier().equals(sender.defaultSchoolDataIdentifier())) {
+      return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    } else {
+      List<CommunicatorMessageRecipientUserGroup> userGroupRecipientsList = communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+      List<CommunicatorMessageRecipientUserGroup> userGroupRecipients = new ArrayList<CommunicatorMessageRecipientUserGroup>();
+      SchoolDataIdentifier loggedUserIdentifier = sessionController.getLoggedUserEntity().defaultSchoolDataIdentifier();
+  
+      for (CommunicatorMessageRecipientUserGroup userGroup : userGroupRecipientsList) {
+        UserGroupEntity userGroupEntity = userGroupEntityController.findUserGroupEntityById(userGroup.getUserGroupEntityId());
+        List<UserGroupUserEntity> groupUsers = userGroupEntityController.listUserGroupUserEntitiesByUserGroupEntity(userGroupEntity);
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java:371]: Might be faster using `userGroupEntityController.isMember(loggedUserIdentifier, userGroupEntity)`

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java [hunk 2]
```diff
@@ -350,11 +355,50 @@ public CommunicatorMessageRecipient findCommunicatorMessageRecipientByMessageAnd
...
-    return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    UserEntity loggedUser = sessionController.getLoggedUserEntity();
+    UserEntity sender = userEntityController.findUserEntityById(communicatorMessage.getSender());
+    if (sessionController.hasEnvironmentPermission(CommunicatorPermissionCollection.COMMUNICATOR_READ_RECIPIENTS_LIST)) {
+      return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    } else if (loggedUser.defaultSchoolDataIdentifier().equals(sender.defaultSchoolDataIdentifier())) {
+      return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    } else {
+      List<CommunicatorMessageRecipientUserGroup> userGroupRecipientsList = communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+      List<CommunicatorMessageRecipientUserGroup> userGroupRecipients = new ArrayList<CommunicatorMessageRecipientUserGroup>();
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java:366]: `List<CommunicatorMessageRecipientUserGroup> userGroupRecipients = new ArrayList<>();`

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java [hunk 3]
```diff
@@ -350,11 +355,50 @@ public CommunicatorMessageRecipient findCommunicatorMessageRecipientByMessageAnd
...
-    return communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+    UserEntity loggedUser = sessionController.getLoggedUserEntity();
+    UserEntity sender = userEntityController.findUserEntityById(communicatorMessage.getSender());
+    if (sessionController.hasEnvironmentPermission(CommunicatorPermissionCollection.COMMUNICATOR_READ_RECIPIENTS_LIST)) {
+      return communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+    } else if (loggedUser.defaultSchoolDataIdentifier().equals(sender.defaultSchoolDataIdentifier())) {
+      return communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+    } else {
+      List<CommunicatorMessageRecipientWorkspaceGroup> workspaceGroupRecipientsList = communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+      List<CommunicatorMessageRecipientWorkspaceGroup> workspaceGroupRecipients = new ArrayList<CommunicatorMessageRecipientWorkspaceGroup>();
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java:390]: `List<CommunicatorMessageRecipientWorkspaceGroup> workspaceGroupRecipients = new ArrayList<>();`

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java [hunk 4]
```diff
@@ -350,11 +355,50 @@ public CommunicatorMessageRecipient findCommunicatorMessageRecipientByMessageAnd
...
+      for (CommunicatorMessageRecipientWorkspaceGroup workspaceGroup : workspaceGroupRecipientsList) {
+        UserEntity loggedUserEntity = sessionController.getLoggedUserEntity();
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java:392]: Move outside the `for` loop; will be the same for each iteration, anyway.

**Review by pkukkon — CHANGES_REQUESTED** [review]

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java (hunk 1 above)
Comment (anttiviljakainen) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java:371]: Seconded, the `groupUsers.contains(groupUserEntity)` below can be bit iffy as it's going to use `object.equals`, which I wouldn't count on when the objects are Entities that don't have explicit compare functions.

**Review by anttiviljakainen — CHANGES_REQUESTED** [review]

**Review by pkukkon — DISMISSED** [review]

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java [hunk 5]
```diff
@@ -350,11 +355,51 @@ public CommunicatorMessageRecipient findCommunicatorMessageRecipientByMessageAnd
...
+    UserEntity loggedUser = sessionController.getLoggedUserEntity();
+    UserEntity sender = userEntityController.findUserEntityById(communicatorMessage.getSender());
+    if (sessionController.hasEnvironmentPermission(CommunicatorPermissionCollection.COMMUNICATOR_READ_RECIPIENTS_LIST)) {
+      return communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+    } else if (loggedUser.defaultSchoolDataIdentifier().equals(sender.defaultSchoolDataIdentifier())) {
+      return communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+    } else {
+      List<CommunicatorMessageRecipientWorkspaceGroup> workspaceGroupRecipientsList = communicatorMessageRecipientWorkspaceGroupDAO.listByMessage(communicatorMessage);
+      List<CommunicatorMessageRecipientWorkspaceGroup> workspaceGroupRecipients = new ArrayList<>();
+      UserEntity loggedUserEntity = sessionController.getLoggedUserEntity();
```
Comment (anttiviljakainen) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java:391]: There's already a loggedUser above

**Review by anttiviljakainen — APPROVED** [review]

**Review by pkukkon — APPROVED** [review]

**Pull request merged**